README.md

# 📄 swift-web-page (swep)

<p align="left">
  <img src="https://img.shields.io/badge/Swift_Version-5.1-orange.svg?style=flat&logo=Swift" alt="Swift Version: 5.1"/>
  <a href="https://swift.org/package-manager">
    <img src="https://img.shields.io/badge/SwiftPM-Compatible-darkgreen.svg?style=flat" alt="Swift Package Manager"/>
  </a>
    <img src="https://img.shields.io/badge/Platforms-macOS,%20iOS,%20Linux-darkgreen.svg?style=flat" alt="Swift Package Manager"/>
</p>

Swep is a Swift DSL for writing type-safe HTML/CSS in SwiftUI way.

## Table of Contents

  - [Motivation](#motivation)
  - [Examples](#examples)
  - [Safety](#safety)
  - [Design](#design)
  - [FAQ](#faq)
  - [Installation](#installation)
  - [Real world example](#real-world-example)
  - [License](#license)

## Motivation

The popular choice for rendering HTML/CSS in Swift these days is to use templating languages, but they expose your application to **runtime errors** and **invalid HTML/CSS**. Our library prevents these runtime issues at compile-time by embedding HTML/CSS directly into Swift’s powerful type system.

## Examples

HTML/CSS documents can be created in a SwiftUI-like fashion, much like you might create a nested SwiftUI View:

```swift
import Swep

let page = document {
  html {
    head(title("YOLO!"))
    body {
      h1("Welcome!")
      p("You've found our site!")
    }
  }
}
```

CSS inside Style tag!

```swift
import Swep

let page = document {
  html {
    head {
      title("YOLO!")
      style {
        selector("h1") {
          color(.tomato())
        }
      }
    }
    body {
      h1("Welcome!")
      p("You've found our site!")
    }
  }
}
```

Even better, CSS as tag modifier

```swift
import Swep

let page = document {
  html {
    head(title("YOLO!"))
    body {
      h1("Welcome!")
        .color(.tomato())
      p("You've found our site!")
    }
  }
}
```

Once your document is created you can render it using the `render` function:

```swift
/// when you just want to see the output..
page.render(.debug(.pretty(.spaces(2))))
// or
page.debugRender()
```

```html
<!DOCTYPE html>
<html>
  <head>
    <title>
      YOLO!
    </title>
  </head>
  <body>
    <h1 style="color:rgba(255,99,71,1.0)">
      Welcome!
    </h1>
    <p>
      You've found our site!
    </p>
  </body>
</html>
```

```swift
/// when your document is ready for release..
page.render(.release(.inline))
// or
page.render()
```

```html
<!DOCTYPE html><html><head><title>YOLO!</title></head><body><h1 style="color:rgba(255,99,71,1.0);">Welcome!</h1><p>You&apos;ve found our site!</p></body></html>
```

## Safety

Because we are embedding our DSL in Swift we can take advantage of some advanced Swift features to add an extra layer of safety when constructing HTML/CSS documents. For a simple example, we can strengthen many HTML/CSS APIs to force their true types rather than just relying on strings. 

```swift 
let imgTag = img()
               .src("cat.jpg")
               .width(400)
               .height(300)

imgTag.render()
// <img src="cat.jpg" width="400" height="300">
```

Here the `src` attribute takes a string, but `width` and `height` take integers, as it’s invalid to put anything else in those attributes.

For a more advanced example, `<li>` tags can only be placed inside `<ol>` and `<ul>` tags, and we can represent this fact so that it’s impossible to construct an invalid document:

```swift
let listTag = ul {
  li("Cat")
  li("Dog")
  li("Rabbit")
} // ✅ Compiles!

listTag.render()
// <ul><li>Cat</li><li>Dog</li><li>Rabbit</li></ul>

div {
  li("Cat")
  li("Dog")
  li("Rabbit")
} // 🛑 Compile error
```

Another advance example, `<thead>`, `<tbody>`, and `<tfoot>` tags can only be placed inside `<table>` tags, same for `<th>`, and `<td>` tags can only be placed inside `<tr>` tags.

```swift
let tableTag = table {
  tr(th("A Head"))
  tr(td("A Body"))
} // ✅ Compiles!

type(of: tableTag)
// Table<Tuple<(Tr<Th<Text>>, Tr<Td<Text>>)>>

let tableTag = table {
  thead(tr(th("A Head")))
  tbody(tr(td("A Body")))
} // ✅ Compiles!

type(of: tableTag)
// Table<Tuple<(Thead<Tr<Th<Text>>>, Tbody<Tr<Td<Text>>>)>>

let tableTag = table {
  thead(tr(th("A Head")))
  for _ in 1...3 {
    tbody(tr(td("A Body")))
  }
} // ✅ Compiles!

type(of: tableTag)
// Table<Tuple<(Thead<Tr<Th<Text>>>, Array<Tbody<Tr<Td<Text>>>>)>>

table {
  tbody(tr(td("A Body")))
  thead(tr(th("A Head")))
} // 🛑 Compile error
```

There alot more..

## Design

Behind the scenes **Swep** is following the **Protocol-Oriented-Programming** (**POP**) approach, along with the powerful swift-feature **[@resultBuilder](https://github.com/apple/swift-evolution/blob/main/proposals/0289-result-builders.md)**. There are two main-libraries *Html* handles the html-side, and *Css* handles the css-side. And one micro-library *HtmlCssSupport* which handles the combination of both *Html*, and *Css*.

## FAQ

### Can I use this with existing Swift web frameworks like Vapor, Kitura, and Perfect?

Yes! We even provide plug-in libraries that reduce the friction of using this library with Vapor, Kitura, and Perfect. Find out more information at the following repos:

- [swift-web-page-vapor](https://github.com/alja7dali/swift-web-page-vapor)
- [swift-web-page-kitura](https://github.com/alja7dali/swift-web-page-kitura)
- [swift-web-page-perfect](https://github.com/alja7dali/swift-web-page-perfect)
 
### Why would I use this over a templating language?

Templating languages are popular and easy to get started with, but they have many drawbacks:

1. **Stringy APIs**: Templating languages are always stringly typed because you provide your template as a big ole string, and then at runtime the values are interpolated and logic is executed. This means things we take for granted in Swift, like the compiler catching typos and type mismatches, will go unnoticed until you run the code.

1. **Incomplete language**: Templating languages are just that: programming languages. That means you should expect from these languages all of the niceties you get from other fully-fledged languages like Swift. That includes syntax highlighting, IDE autocompletion, static analysis, refactoring tools, breakpoints, debugger, and a whole slew of features that make Swift powerful like let-bindings, conditionals, loops and more. However, the reality is that no templating language supports all of these features.

1. **Rigid**: Templating languages are rigid in that they do not allow the types of compositions and transformations we are used to performing on data structures in Swift. It is not possible to succinctly traverse over the documents you build, and inspect or transform the nodes you visit. This capability has many applications, such as being able to pretty print or minify your HTML/CSS output, or writing a transformation that allows you to inline a CSS stylesheet into an HTML/CSS node. There are entire worlds closed off to you due to how templating languages work.

The DSL in this library fixes all of these problems, and opens up doors that are completely closed to templating languages.

### When is it more appropriate to use a templating language over swift-web-page?

There are a few reasons you might want to still use a templating language:

1. A designer delivers a large HTML/CSS document to you and all you want to do is hook in a little bit of value interpolation or logic. In this case you can simply copy and paste that HTML/CSS into your template, add a few interpolation tokens, and you're well on your way to having a full page served from your web application.

1. You need to render non-HTML/CSS documents. The beauty of templating languages is that it outputs straight to plain text, and so it can model any type of document, whether it be HTML/CSS, markdown, XML, RSS, ATOM, LaTeX, and more.

1. Creating _very_ large documents in a single expression can cause compile times to go up, whereas templates are not compiled by Swift and so do not influence compile times. Luckily this isn't a problem too often because it is very easy to break up a document into as many small pieces as you want, which will probably lead to more reusable code in the long run.

If you do decide that a templating language better suites your needs, then you should consider [HypertextLiteral](https://github.com/NSHipster/HypertextLiteral), which gives you template-like capabilities but in a safer manner.

## Real world example

Creating a html document, and css stylesheet

```swift
let titillimFont: StaticString = """
  https://fonts.googleapis.com/css2?family=\
  Titillium+Web:ital,wght@0,200;0,300;0,400;\
  0,600;0,700;0,900;1,200;1,300;1,400;1,600;\
  1,700&display=swap
  """
  
let plainStyle = style {
  `import`(titillimFont)
  selector("*, *::before, *::after") {
    margin(.px(0))
    padding(.px(0))
    boxSizing(.borderBox)
  }
  selector("body") {
    margin(.px(0), .auto)
    backgroundColor(.hex(0x111))
    fontFamily("'Titillium Web', sans-serif")
  }
}

let page = document {
  html {
    head {
      title("Hello, Swep!")
      plainStyle
    }
    body {
      h1("📄 swift-web-page (swep)")
      hr()
        .width(50%)
        .minWidth(.px(720))
        .color(.hex(0x323232))
      p {
        strong("Swep ")
        text("is a Swift DSL for writing type-safe HTML/CSS in declarative way.")
      }
      p {
        text("With ")
        strong("Swep:")
      }
      ul {
        li("You can write html documents along with css")
          .fontWeight(.bolder)
        li("Bringing all swift-language features out of the box")
        #if swift(>=5.1)
          for version in 1...4 {
            if version != 2 {
              li("supporting swift v5.\(version)")
            }
          }
        #else
          li("Unfortunately this library is built using @resultBuilder which is available in swift v5.1 and higher 😢")
        #endif
      }
      blockquote("Enjoy! ✌️😁")
    }
  }
}
```

Checking out the type of the document..
```swift
type(of: page.content)
```

```bash
Html<Tuple<(Head<Tuple<(Title<Text>, Style<Stylesheet>)>>, Body<Tuple<(H1<Text>, Hr, P<Tuple<(Strong<Text>, Text)>>, P<Tuple<(Text, Strong<Text>)>>, Ul<Tuple<(Li<Text>, Li<Text>, Array<Optional<Li<Text>>>)>>, Blockquote<Text>)>>)>>
```

Rendering out the document..

```swift
let renderMode: RenderMode = .release(.pretty(.spaces(2)))
page.render(renderMode)
```

```html
<!DOCTYPE html>
<html>
  <head>
    <title>
      Hello, Swep!
    </title>
    <style>

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

<hr align="center" style="background-color: #393939; height: 2px">

# Releases
  - [0.0.1](#001)

<hr align="center" style="background-color: #393939; height: 2px">

## 0.0.1
> **Swep** initial release

Sources/main/main.swift
// // import Swep

// // // let page = document {
// // //   html {
// // //     head {
// // //       title("YOLO!")
// // //     }
// // //     body {
// // //       h1("Welcome!")
// // //       p("You've found our site!")
// // //     }
// // //   }
// // // }

// // // let page = document {
// // //   html {
// // //     head {
// // //       title("YOLO!")
// // //       style {
// // //         selector("h1") {
// // //           color(.tomato())
// // //         }
// // //       }
// // //     }
// // //     body {
// // //       h1("Welcome!")
// // //       p("You've found our site!")
// // //     }
// // //   }
// // // }



// // // let page = document {
// // //   html {
// // //     head(title("YOLO!"))
// // //     body {
// // //       h1("Welcome!")
// // //         .color(.tomato())
// // //       p("You've found our site!")
// // //     }
// // //   }
// // // }

// // // print(page.render(.release(.inline)))

// // // print(page.render(.debug(.pretty(.spaces(2)))))
























// // // func welcomeSection() -> some HtmlNode {
// // //   return div {
// // //     p(text("Hide me pls!"))
// // //       .display(.none)
// // //     h1(text("Welcome!"))
// // //     p("Welcome to you & who has come here before.")
// // //     a("a link")
// // //       .href("https://example.org")
// // //     p {
// // //       text("This is a paragraph with ")
// // //       em("some")
// // //       text(" inline formatting and ")
// // //       a("a link")
// // //         .href("https://example.org")
// // //       text("!")
// // //     }
// // //   }
// // // }

// // // func orderedList() -> some HtmlNode {
// // //   return div {
// // //     p {
// // //       text("Here's an ")
// // //       strong("ordered")
// // //       text(" list:")
// // //     }
// // //     ol {
// // //       li("Preheat oven.")
// // //       li("Mix ingredients.")
// // //       li("Order delivery.")
// // //     }
// // //   }
// // // }

// // // func unorderedList() -> some HtmlNode {
// // //   return div {
// // //     p("And \"unordered\":")
// // //     ol {
// // //       li("Preheat oven.")
// // //       li("Mix ingredients.")
// // //       li("Order delivery.")
// // //     }
// // //   }
// // // }

// // // func quiteOfTheDay() -> some HtmlNode {
// // //   return div {
// // //     p("They said:")
// // //     blockquote("Code to the interface you wish you had not the interface you were given.")
// // //     script("alert(\"hello!\");")
// // //   }
// // // }

// // // let page = document {
// // //   html {
// // //     head {
// // //       title("Hello!")
// // //       style("h1{color: red}")
// // //     }
// // //     body {
// // //       welcomeSection()
// // //       orderedList()
// // //       unorderedList()
// // //       quiteOfTheDay()
// // //     }
// // //   }
// // // }


// // // print(myUnorderedList().debugRender())


// // let myBody = body {
// //   // div {
// //     p(text("Hide me pls!"))
// //       .display(.none)
// //     h1(text("Welcome!"))
// //     p("Welcome to you & who has come here before.")
// //     a("a link")
// //       .href("https://example.org")
// //     p {
// //       text("This is a paragraph with ")
// //       em("some")
// //       text(" inline formatting and ")
// //       a("a link")
// //         .href("https://example.org")
// //       text("!")
// //     }
// //   }
// //   // div {
// //   //   p {
// //   //     text("Here's an ")
// //   //     strong("ordered")
// //   //     text(" list:")
// //   //   }
// //   //   ol {
// //   //     li("Preheat oven.")
// //   //     li("Mix ingredients.")
// //   //     li("Order delivery.")
// //   //   }
// //   // }
// // //   div {
// // //     p("And \"unordered\":")
// // //     ol {
// // //       li("Preheat oven.")
// // //       li("Mix ingredients.")
// // //       li("Order delivery.")
// // //     }
// // //   }
// // //   div {
// // //     p("They said:")
// // //     blockquote("Code to the interface you wish you had not the interface you were given.")
// // //     script("alert(\"hello!\");")
// // //   }
// // // }

// // print(type(of: myBody.self))




// // // // print(HtmlNodeBuilder.buildOptional(Text("Hello")))

// // // let nilText: Optional<Text> = .none

// // // let header1 = H1 {
// // //   // nilText

// // //   if true {
// // //     Raw("Stake!")
// // //   }
  
// // //   // // (true) ? Raw("Stake!") : Empty()
// // //   // ([true]).map { $0 ? Raw("Stake!") : Raw("") }
// // // }

// // // let header2 = H1 {
// // //   // nilText

// // //   for _ in 0...10 {
// // //     Text("Stake!")
// // //   }
  
// // //   // (0...10).map { _ in
// // //   //   Raw("Stake!")
// // //   // }

// // // }




// // // let isTrue = true

// // // let header3 = H1 {
// // //   if isTrue {
// // //     if isTrue {
// // //       Text("1")
// // //       // Raw("2")
// // //     } else {
// // //       Raw("2")
// // //     }
// // //   } else {
// // //     if isTrue {
// // //       Text("1")
// // //     } else {
// // //       Raw("2")
// // //       // Text("3")
// // //     }
// // //   }
  
// // //     if isTrue {
// // //       Div {
// // //         Text("1")
// // //         Raw("2")
// // //       }
// // //     } else {
// // //       Raw("2")
// // //     }
  
// // //   // isTrue ? Text("3") : Text("4")
// // // }

// // // print(header1.render())
// // // print(type(of: header1.content))

// // // print(header2.render())
// // // print(type(of: header2.content))

// // // print(header3.render())
// // // print(type(of: header3.content))













// // // fileprivate func childOfBuilderGenerator() {
// // //   let childOf = [
// // //     "Any",
// // //     "Audio",
// // //     "Colgroup",
// // //     "Details",
// // //     "Dl",
// // //     "Fieldset",
// // //     "Figure",
// // //     "Head",
// // //     "Map",
// // //     "Object",
// // //     "Ol",
// // //     "Optgroup",
// // //     "Picture",
// // //     "Select",
// // //     // "Table",
// // //     "Tbody",
// // //     "Tfoot",
// // //     "Thead",
// // //     // "Tr",
// // //     "Ul",
// // //     "Video"
// // //   ]
  
// // //   let min = 1, max = 10
// // //     print("public extension HtmlNodeBuilder {")
// // //   for i in min..<max {
// // //     var generic = ""
// // //     var parameters = ""
// // //     var returnType = ""
// // //     var returnValue = ""
// // //     for j in 0...i {
// // //       generic += "T\(j): HtmlNode, "
// // //       parameters += "_ v\(j): T\(j), "
// // //       returnType += "T\(j), "
// // //       returnValue += "v\(j), "
// // //     }
// // //     print("  @_disfavoredOverload")
// // //     print("  static func buildBlock<\(generic.dropLast(2))>(")
// // //     print("    \(parameters.dropLast(2))")
// // //     print("  ) -> Tuple<(\(returnType.dropLast(2)))> {")
// // //     print("    return Tuple<(\(returnType.dropLast(2)))>( (\(returnValue.dropLast(2))) )")
// // //     print("  }")
// // //   }
// // //     print("}")
// // //     print()

// // //   for c in childOf {
// // //       print("/// MARK: ChildOf\(c)")
// // //       print("public typealias ChildOf\(c)Node = ChildOf\(c) & HtmlNode")
// // //       print("public extension Element where Kind: ChildOf\(c) {")
// // //     for i in min..<max {
// // //       var generic = ""
// // //       var parameters = ""
// // //       var returnType = ""
// // //       var returnValue = ""
// // //       for j in 0...i {
// // //         generic += "T\(j): ChildOf\(c)Node, "
// // //         parameters += "_ v\(j): T\(j), "
// // //         returnType += "T\(j), "
// // //         returnValue += "v\(j), "
// // //       }
// // //       // print("  @_disfavoredOverload")
// // //       print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Tuple<(\(returnType.dropLast(2)))>) { self.init([], .content()) }")
// // //     }
// // //       print("}")
// // //       print()
// // //   }
// // // }

// // // childOfBuilderGenerator()






// // // let renderMode: RenderMode = .release(.pretty(.spaces(2)))

// // // let darkGithubStyle = style(renderMode) {
// // //   selector("body") {
// // //     margin(.px(0), .auto)
// // //     backgroundColor(.black())
// // //     fontFamily("Georgia, .Palatino, .serif")
// // //     color(.hex(0xEEE))
// // //     lineHeight(1)
// // //     maxWidth(.px(960))
// // //     padding(.px(30))
// // //   }
// // //   selector("h1, .h2, .h3, .h4") {
// // //     fontWeight(400)
// // //   }
// // //   selector("h1, .h2, .h3, .h4, .h5, p") {
// // //     marginBottom(.px(24))
// // //     padding(.px(0))
// // //   }
// // //   selector("h1") {
// // //     fontSize(.px(48))
// // //   }
// // //   selector("h2") {
// // //     fontSize(.px(36))
// // //     margin(.px(24), .px(0), .px(6))
// // //   }
// // //   selector("h3") {
// // //     fontSize(.px(24))
// // //   }
// // //   selector("h4") {
// // //     fontSize(.px(21))
// // //   }
// // //   selector("h5") {
// // //     fontSize(.px(18))
// // //   }
// // //   selector("a") {
// // //     color(.hex(0x61BFC1))
// // //     margin(.px(0))
// // //     padding(.px(0))
// // //     textDecorationLine(.none)
// // //     // verticalAlignment(.underline)
// // //   }
// // //   selector("a:hover") {
// // //     textDecorationLine(.underline)
// // //   }
// // //   selector("a:visited") {
// // //     color(.hex(0x466B6C))
// // //   }
// // //   selector("ul, .ol") {
// // //     margin(.px(0))
// // //     padding(.px(0))
// // //   }
// // //   selector("li") {
// // //     lineHeight(.px(24))
// // //   }
// // //   selector("li ul, .li ul") {
// // //     marginLeft(.px(24))
// // //   }
// // //   selector("p, .ul, .ol") {
// // //     fontSize(.px(16))
// // //     lineHeight(.px(24))
// // //     maxWidth(.px(540))
// // //   }
// // //   selector("pre") {
// // //     padding(.px(0), .px(24))
// // //     maxWidth(.px(800))
// // //     // whiteSpace(.prewrap)
// // //   }
// // //   selector("code") {
// // //     fontFamily("Consolas, .Monaco, .Andale Mono, .monospace")
// // //     lineHeight(1.5)
// // //     fontSize(.px(13))
// // //   }
// // //   selector("aside") {
// // //     display(.block)
// // //     float(.right)
// // //     width(.px(390))
// // //   }
// // //   selector("blockquote") {
// // //     borderLeft(
// // //       width: .em(5),
// // //       style: .solid,
// // //       color: .hex(0xEEE)
// // //     )
// // //     padding(.px(0), .em(2))
// // //     marginLeft(.px(0))
// // //     maxWidth(.px(476))
// // //   }
// // //   selector("blockquote cite") {
// // //     fontSize(.px(14))
// // //     lineHeight(.px(20))
// // //     color(.hex(0xBFBFBF))
// // //   }
// // //   selector("blockquote cite:before") {
// // //     content("\'\\2014 \\00A0\'")
// // //   }
// // //   selector("blockquote p") {
// // //     color(.hex(0x666))
// // //     maxWidth(.px(460))
// // //   }
// // //   selector("hr") {
// // //     width(.px(540))
// // //     textAlign(.left)
// // //     margin(.px(0), .auto, .px(0), .px(0))
// // //     color(.hex(0x999))
// // //   }
// // // }

// // // let page = document {
// // //   html {
// // //     head {
// // //       title("Hello!")
// // //       darkGithubStyle
// // //     }
// // //     body {
// // //       p("Hide me pls!")
// // //         .display(.none)
// // //       h1("Welcome!")
// // //       p("Welcome to you & who has come here before.")
// // //       p {
// // //         text("This is a paragraph with ")
// // //         em("some")
// // //         text(" inline formatting and ")
// // //         a("a link")
// // //           .href("https://example.org")
// // //         text("!")
// // //       }
// // //       p {
// // //         text("Here's an ")
// // //         strong("ordered")
// // //         text(" list:")
// // //       }
// // //       ol {
// // //         li("Eat")
// // //         li("Code")
// // //         li("Sleep")
// // //         li("Repeat")
// // //       }
// // //       p("And \"unordered\":")
// // //       ul {
// // //         li("Milk")
// // //         li("Eggs")
// // //         li("Coffee")
// // //       }
// // //       p("They said:")
// // //       blockquote("Code to the interface you wish you had not the interface you were given.")
// // //       script("alert(\"hello!\");")
// // //     }
// // //   }
// // // }

// // // print(page.render(renderMode))





// import Swep

// // let renderMode: RenderMode = .release(.pretty(.spaces(2)))

// // let myDiv1 = div {
// //   p("Hide me pls!")
// //     // .display(.none)
// //   h1("Welcome!")
// //   p("Welcome to you & who has come here before.")
// //   p {
// //     text("This is a paragraph with ")
// //     em("some")
// //     text(" inline formatting and ")
// //     a("a link")
// //       // .href("https://example.org")
// //     text("!")
// //   }
// // }
// // let myDiv2 = div {
// //   p {
// //     text("Here's an ")
// //     strong("ordered")
// //     text(" list:")
// //   }
// //   ol {
// //     li("Eat")
// //     li("Code")
// //     li("Sleep")
// //     li("Repeat")
// //   }
// // }

// // func unorderedList() -> some Element {
// //   div {
// //     p("And \"unordered\":")
// //     ul {
// //       li("Milk")
// //       li("Eggs")
// //       li("Coffee")
// //     }
// //     p("They said:")
// //     blockquote("Code to the interface you wish you had not the interface you were given.")
// //     script("alert(\"hello!\");")
// //   }
// // }

// // let page = 
// // document {
// //   html {
// //     head {
// //       title("Hello!")
// //       // darkGithubStyle
// //     }
// //     body {
// //       myDiv1
// //       myDiv2
// //       // unorderedList()
// //     }
// //   }
// // }

// // print(type(of: page.self))
// // print(page.render(renderMode))


// // func someText() -> some Tag {
// //   p("Hello, .world!")
// // }

// // // let h = h1 {
// // //   p("Hello, .world!")
// // //   someText()
// // // }

// // print(type(of: someText().self))
// // print(someText().render(renderMode))





// // let renderMode: RenderMode = .release(.pretty(.spaces(2)))

// // let titillimFont: StaticString = """
// //   https://fonts.googleapis.com/css2?family=\
// //   Titillium+Web:ital,wght@0,200;0,300;0,400;\
// //   0,600;0,700;0,900;1,200;1,300;1,400;1,600;\
// //   1,700&display=swap
// //   """
  
// // let plainStyle = style {
// //   `import`(titillimFont)
// //   selector("*, *::before, *::after") {
// //     margin(.px(0))
// //     padding(.px(0))
// //     boxSizing(.borderBox)
// //   }
// //   selector("body") {
// //     margin(.px(0), .auto)
// //     backgroundColor(.hex(0x111))
// //     fontFamily("'Titillium Web', .sans-serif")
// //   }
// // }

// // let page = document {
// //   html {
// //     head {
// //       title("Hello, .Swep!")
// //       plainStyle
// //     }
// //     body {
// //       h1("📄 swift-web-page (swep)")
// //       hr()
// //         .width(50%)
// //         .minWidth(.px(720))
// //         .color(.hex(0x323232))
// //       p {
// //         strong("Swep ")
// //         text("is a Swift DSL for writing type-safe HTML/CSS in declarative way.")
// //       }
// //       p {
// //         text("With ")
// //         strong("Swep:")
// //       }
// //       ul {
// //         li("You can write html documents along with css")
// //           .fontWeight(.bolder)
// //         li("Bringing all swift-language features out of the box")
// //         #if swift(>=5.1)
// //           for version in 1...4 {
// //             if version != 2 {
// //               li("supporting swift v5.\(version)")
// //             }
// //           }
// //         #else
// //           li("Unfortunatly this library is built using @resultBuilder which available in swift v5.1 and higher 😢")
// //         #endif
// //       }
// //       blockquote("Enjoy! ✌️😁")
// //     }
// //   }
// // }

// // print(type(of: page))
// // print(page.render(renderMode))




// // let tableTag = table {
// //   tr(th("A Head"))
// //   tr(td("A Body"))
// // } // ✅ Compiles!

// // print(type(of: tableTag))

// // let theadTag = thead(tr(th("A Head")))
// // let tbodyTag = tbody(tr(td("A Body")))

// // let tableTag = table {
// //   theadTag
// //   tbodyTag
// // } // ✅ Compiles!

// // print(type(of: tableTag))

// let theadTag = thead(tr(th("A Head")))
// let tbodyTag = (1...3).map { _ in
//   tbody(tr(td("A Body")))
// }

// let tableTag = table {
//   theadTag
//   tbodyTag
  
// } // ✅ Compiles!

// print(type(of: tableTag))














// // /// Audio
// // let myAudio = Audio {
// //   AudioSource()
// //   Track(src: "I have more info")
// // }

// // print(type(of: myAudio.self))

// // /// Colgroup
// // let myColgroup = Colgroup {
// //   for _ in 0..<10 {
// //     Col()
// //   }
// // }

// // print(type(of: myColgroup.self))

// // /// Details
// // let myDetails = Details {
// //   Summary("Info")
// //   P("I have more info")
// // }

// // print(type(of: myDetails.self))

// // /// Dl
// // let myDl = Dl {
// //   Dd("Info")
// //   Dt("I have more info")
// // }

// // print(type(of: myDl.self))

// // /// Fieldset
// // let myFieldset = Fieldset {
// //   Legend("Info")
// //   P("I have more info")
// // }

// // print(type(of: myFieldset.self))

// // /// Figure
// // let myFigure = Figure {
// //   Figcaption("Info")
// //   P("I have more info")
// // }

// // print(type(of: myFigure.self))


// // /// TrTh
// // let myTrTh = Tr {
// //   Th("Info")
// //   Th("I have more info")
// // }

// // print(type(of: myTrTh.self))

// // /// TrTd
// // let myTrTd = Tr {
// //   Td("Info")
// //   Td("I have more info")
// // }

// // print(type(of: myTrTd.self))

// // /// Thead
// // let myThead = Thead {
// //   Tr {
// //     Th("Info")
// //     Th("I have more info")
// //   }
// // }

// // print(type(of: myThead.self))

// // /// Tbody
// // let myTbody = Tbody {
// //   Tr {
// //     Td("Info")
// //     Td("I have more info")
// //   }
// // }

// // print(type(of: myTbody.self))

// // /// Tfoot
// // let myTfoot = Tfoot {
// //   Tr {
// //     Td("Info")
// //     Td("I have more info")
// //   }
// // }

// // print(type(of: myTfoot.self))





















// // for i in 1..<10 {
// //   var generic = ""
// //   var parameters = ""
// //   var returnType = ""
// //   var returnValue = ""
// //     generic += "T0: ChildOfThead, "
// //     parameters += "_ v0: T0, "
// //     returnType += "T0, "
// //     returnValue += "v0, "
// //   for j in 1...i {
// //     generic += "T\(j): VariationOfTbody, "
// //     parameters += "_ v\(j): T\(j), "
// //     returnType += "T\(j), "
// //     returnValue += "v\(j), "
// //   }
// //   print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Content) where Content == Tuple<(\(returnType.dropLast(2)))> { self.init([], .content()) }")
// // }

// // for i in 1..<10 {
// //   var generic = ""
// //   var parameters = ""
// //   var returnType = ""
// //   var returnValue = ""
// //     generic += "T0: ChildOfThead, "
// //     parameters += "_ v0: Thead<T0>, "
// //     returnType += "Thead<T0>, "
// //     returnValue += "v0, "
// //   for j in 1..<i {
// //     generic += "T\(j): VariationOfTbody, "
// //     parameters += "_ v\(j): Tbody<T\(j)>, "
// //     returnType += "T\(j), "
// //     returnValue += "v\(j), "
// //   }
// //     generic += "T\(i): ChildOfTfoot, "
// //     parameters += "_ v\(i): Tfoot<T\(i)>, "
// //     returnType += "Tfoot<T\(i)>, "
// //     returnValue += "v\(i), "
// //   print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Content) where Content == Tuple<(\(returnType.dropLast(2)))> { self.init([], .content()) }")
// // }


// //   /// Caption, Thead, Tbody ...
// //   /// Colgroup, Thead, Tbody ...
// //   /// Caption, Colgroup, Thead, Tbody ...

// // for i in 2..<10 {
// //   var generic = ""
// //   var parameters = ""
// //   var returnType = ""
// //   var returnValue = ""
// //     generic += "T0: ChildOfAny, "
// //     parameters += "_ v0: Caption<T0>, "
// //     returnType += "Caption<T0>, "
// //     returnValue += "v0, "
// //     generic += "T1: ChildOfThead, "
// //     parameters += "_ v1: Thead<T1>, "
// //     returnType += "Thead<T1>, "
// //     returnValue += "v1, "
// //   for j in 2...i {
// //     if j < i {
// //       generic += "T\(j): VariationOfTbody, "
// //       parameters += "_ v\(j): T\(j), "
// //       returnType += "T\(j), "
// //       returnValue += "v\(j), "
// //     } else {
// //       generic += "T\(j): ChildOfTfoot, "
// //       parameters += "_ v\(j): Tfoot<T\(j)>, "
// //       returnType += "Tfoot<T\(j)>, "
// //       returnValue += "v\(j), "
// //     }
// //   }
// //   print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Content) where Content == Tuple<(\(returnType.dropLast(2)))> { self.init([], .content()) }")
// // }

// // for i in 2..<10 {
// //   var generic = ""
// //   var parameters = ""
// //   var returnType = ""
// //   var returnValue = ""
// //     generic += "T0: ChildOfColgroup, "
// //     parameters += "_ v0: Colgroup<T0>, "
// //     returnType += "Colgroup<T0>, "
// //     returnValue += "v0, "
// //     generic += "T1: ChildOfThead, "
// //     parameters += "_ v1: Thead<T1>, "
// //     returnType += "Thead<T1>, "
// //     returnValue += "v1, "
// //   for j in 2...i {
// //     if j < i {
// //       generic += "T\(j): VariationOfTbody, "
// //       parameters += "_ v\(j): T\(j), "
// //       returnType += "T\(j), "
// //       returnValue += "v\(j), "
// //     } else {
// //       generic += "T\(j): ChildOfTfoot, "
// //       parameters += "_ v\(j): Tfoot<T\(j)>, "
// //       returnType += "Tfoot<T\(j)>, "
// //       returnValue += "v\(j), "
// //     }
// //   }
// //   print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Content) where Content == Tuple<(\(returnType.dropLast(2)))> { self.init([], .content()) }")
// // }

// // for i in 3..<10 {
// //   var generic = ""
// //   var parameters = ""
// //   var returnType = ""
// //   var returnValue = ""
// //     generic += "T0: ChildOfAny, "
// //     parameters += "_ v0: Caption<T0>, "
// //     returnType += "Caption<T0>, "
// //     returnValue += "v0, "
// //     generic += "T1: ChildOfColgroup, "
// //     parameters += "_ v0: Colgroup<T1>, "
// //     returnType += "Colgroup<T1>, "
// //     returnValue += "v1, "
// //     generic += "T2: ChildOfThead, "
// //     parameters += "_ v2: Thead<T2>, "
// //     returnType += "Thead<T2>, "
// //     returnValue += "v2, "
// //   for j in 3...i {
// //     if j < i {
// //       generic += "T\(j): VariationOfTbody, "
// //       parameters += "_ v\(j): T\(j), "
// //       returnType += "T\(j), "
// //       returnValue += "v\(j), "
// //     } else {
// //       generic += "T\(j): ChildOfTfoot, "
// //       parameters += "_ v\(j): Tfoot<T\(j)>, "
// //       returnType += "Tfoot<T\(j)>, "
// //       returnValue += "v\(j), "
// //     }
// //   }
// //   print("  init<\(generic.dropLast(2))>(@HtmlNodeBuilder _ content: () -> Content) where Content == Tuple<(\(returnType.dropLast(2)))> { self.init([], .content()) }")
// // }

// //   /// Caption, Thead, Tbody ..., Tfoot
// //   /// Colgroup, Thead, Tbody ..., Tfoot
// //   /// Caption, Colgroup, Thead, Tbody ..., Tfoot






